modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -223,67 +203,78 @@
     ) throws ClientException, ClientError {
         CompletableFuture<T> fut = new CompletableFuture<>();
 
-        ClientChannel ch = channel();
-
-        ch.serviceAsync(op, payloadWriter, payloadReader).handle((res, err) ->
-                handleServiceAsync(op, payloadWriter, payloadReader, fut, null, null, ch, res, err));
+        // Use the only one attempt to avoid blocking async method.
+        handleServiceAsync(fut, op, payloadWriter, payloadReader, 1, null);
 
         return new IgniteClientFutureImpl<>(fut);
     }
 
     /**
      * Handles serviceAsync results and retries as needed.
      */
-    private <T> Object handleServiceAsync(ClientOperation op,
-                                          Consumer<PayloadOutputChannel> payloadWriter,
-                                          Function<PayloadInputChannel, T> payloadReader,
-                                          CompletableFuture<T> fut,
-                                          ClientConnectionException failure,
-                                          AtomicInteger chIdx,
-                                          ClientChannel ch,
-                                          T res,
-                                          Throwable err) {
-        if (err == null) {
-            fut.complete(res);
-            return null;
+    private <T> void handleServiceAsync(final CompletableFuture<T> fut,
+                                        ClientOperation op,
+                                        Consumer<PayloadOutputChannel> payloadWriter,
+                                        Function<PayloadInputChannel, T> payloadReader,
+                                        int attemptsLimit,
+                                        ClientConnectionException failure) {
+        T2<ClientChannel, Integer> chAndAttempts;
+
+        try {
+            chAndAttempts = applyOnDefaultChannel(channel -> channel, attemptsLimit);
```

Comment:
Currently master branch do work approach you describe, and it leads to:
1. introducing additional AtomicInteger variable for attempts counter. Counter is required as it used in separate threads execute `handleServiceAsync` method;
2. code duplication for iterating through channels (for sync and async methods).

This is reasons to use T2. It's not pretty, but alternative isn't looks better, I think.

Reply:
You already get rid of AtomicInteger, and it's not required here anymore. Yes, there will be a little code duplication for iterating through channels, I think it's ok since it will make code more readable.

Reply:
I tried rewrite this code, but it only becomes more complicated (lambdas starts to complain about non-final scope, etc.). S, I suggest another approach with introducing a callback to store attempts counter in an external value (push commit). WDYT?